### PR TITLE
Issue #4349 - test client negotiation of internal extensions

### PR DESCRIPTION
**Pull request by lachlan-roberts**
**Issue #4349**

Added tests for WebSocket client negotiation of requested internal extensions, and added comments to describe the algorithm used to choose the ordering.

Noticed some issues with the parsing of `ExtensionConfig` so fixed that as well.

### CODE REVIEW

File: jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/ExtensionConfig.java [hunk 1]
```diff
@@ -161,17 +162,9 @@ public ExtensionConfig(String name, Map<String, String> parameters)
...
-
-        if (keys.size() > 1)
-            throw new IllegalStateException("parameterizedName contains multiple ExtensionConfigs: " + parameterizedName);
-        if (keys.isEmpty())
-            throw new IllegalStateException("parameterizedName contains no ExtensionConfigs: " + parameterizedName);
-
-        this.name = keys.get(0);
-        this.parameters = new HashMap<>();
-        this.parameters.putAll(paramParser.params.get(this.name));
+        ParamParser paramParser = new ParamParser(parameterizedName).parse();
```
Comment (sbordet) [inline jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/ExtensionConfig.java]: See comment below about usage of the `ParamParser` class.

File: jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/ExtensionConfig.java [hunk 2]
```diff
@@ -343,45 +348,26 @@ else if (paramNameIdx > 0)
...
             String extName = buffer.toString();
-            getParamMap(extName);
+            if (name != null)
+                throw new IllegalArgumentException("parameterizedName contains multiple ExtensionConfigs: " + parameterizedName);
+            name = extName;
             super.parsedValue(buffer);
         }
 
-        private Map<String, String> getParamMap(String extName)
+        public ParamParser parse()
```
Comment (sbordet) [inline jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/ExtensionConfig.java]: Why returning `ParamParser` here? Should not this just be `void`?
The return value is used only once, so perhaps you want to do something different: 
```java
public static ParamParser parse(parameterizedName)
```
The key point is that you are conflating 2 things here: the parsing and the result of the parsing, using the same class for both, which is kind of confusing - hence the `static` method.

File: jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/client/ClientUpgradeRequest.java [hunk 3]
```diff
@@ -298,10 +298,13 @@ public void upgrade(HttpResponse response, HttpConnectionOverHTTP httpConnection
...
+        // Insert client requested internal Extensions into the negotiatedExtension list from the server.
+        // The order of extensions in the response is assumed to be the same as what the client negotiated.
         int i = 0;
         for (ExtensionConfig reqConfig : requestedExtensions)
         {
+            // If this is an internal extension insert in negotiatedExtensions at position i.
+            // If this is not an internal extension we must advance i only if this extension was negotiated by the server.
             if (reqConfig.isInternalExtension())
                 negotiatedExtensions.add(i++, reqConfig);
             else if (i < negotiatedExtensions.size() && reqConfig.getName().equals(negotiatedExtensions.get(i).getName()))
```
Comment (sbordet) [inline jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/client/ClientUpgradeRequest.java]: This is really confusing.

What you want is to remove the extensions that the server did not negotiate from the list of the client extensions (internal and non-internal).

If that's correct, then I think this is more understandable (pseudo-code):

```java
for (Iterator<ExtensionConfig> iterator = clientExtensions.iterator(); iterator.hasNext();) {
  if (!extConfig.isInternal() && !serverExtensions.contains(extConfig)) {
    iterator.remove();
  }
}
```

File: jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/WebSocketNegotiationTest.java [hunk 4]
```diff
@@ -353,4 +359,81 @@ public void testInvalidUpgradeRequestNoKey() throws Exception
...
+        StringBuilder negotiatedExtensions = new StringBuilder();
+        List<Extension> extensions = ((WebSocketCoreSession)clientHandler.coreSession).getExtensionStack().getExtensions();
+        for (int i = 0; i < extensions.size(); i++)
+        {
+            negotiatedExtensions.append(extensions.get(i).getConfig().getParameterizedName());
+            if (i != extensions.size() - 1)
+                negotiatedExtensions.append(", ");
+        }
+        assertThat(negotiatedExtensions.toString(), is(negExts));
+    }
```
Comment (sbordet) [inline jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/WebSocketNegotiationTest.java:431]: This test is fine, but you want to test also the edge cases: just one extension, zero extensions, one internal and one external, etc. - you are always testing for more than 1 extension.

File: jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/WebSocketNegotiationTest.java [hunk 5]
```diff
@@ -353,4 +359,81 @@ public void testInvalidUpgradeRequestNoKey() throws Exception
...
+
+    public static class EmptyExtension extends AbstractExtension
```
Comment (sbordet) [inline jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/WebSocketNegotiationTest.java]: You should not be needing this class. The implementation of the 2 methods can be in `AbstractExtension`, so there is no need to override any method (or override only one when you're only interested in 1 direction).

**Review by sbordet — CHANGES_REQUESTED** [review]

**Comment by sbordet** [comment]
@lachlan-roberts also please avoid in `ClientUpgradeRequest` idioms such as:

```java
if (expression == false)
```

Just use:

```java
if (!expression)
```

**Comment by lachlan-roberts** [comment]
@sbordet these idioms in  `ClientUpgradeRequest` were inherited code from 9.4.x, but I will try to fix things like this if I see it.

**Review by sbordet — APPROVED** [review]
